src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs

```diff
@@ -157,20 +210,43 @@ internal MarshalToken GetToken(IRpcMarshaledContext<object> marshaledContext)
 
         long handle = this.nextUniqueHandle++;
 
+        IRpcMarshaledContext<object> context = JsonRpc.MarshalWithControlledLifetime(declaredType, marshaledObject, options);
+
         IDisposable? revert = this.jsonRpc.AddLocalRpcTargetInternal(
-            marshaledContext.GetType().GenericTypeArguments[0],
-            marshaledContext.Proxy,
-            new JsonRpcTargetOptions(marshaledContext.JsonRpcTargetOptions)
+            declaredType,
+            context.Proxy,
+            new JsonRpcTargetOptions(context.JsonRpcTargetOptions)
             {
                 NotifyClientOfEvents = false, // We don't support this yet.
-                MethodNameTransform = mn => Invariant($"$/invokeProxy/{handle}/{marshaledContext.JsonRpcTargetOptions.MethodNameTransform?.Invoke(mn) ?? mn}"),
+                MethodNameTransform = mn => Invariant($"$/invokeProxy/{handle}/{context.JsonRpcTargetOptions.MethodNameTransform?.Invoke(mn) ?? mn}"),
             },
             requestRevertOption: true);
         Assumes.NotNull(revert);
 
+        Type objectType = marshaledObject.GetType();
+        List<int>? subTypeCodes = null;
+        foreach (RpcMarshalableKnownSubTypeAttribute attribute in GetMarshalableKnownSubtypes(declaredType))
+        {
+            if (attribute.SubType.IsAssignableFrom(objectType))
+            {
+                subTypeCodes ??= new();
+                subTypeCodes.Add(attribute.SubTypeCode);
+
+                this.jsonRpc.AddRpcInterfaceToTargetInternal(
+                    attribute.SubType,
+                    context.Proxy,
+                    new JsonRpcTargetOptions(context.JsonRpcTargetOptions)
+                    {
+                        NotifyClientOfEvents = false, // We don't support this yet.
+                        MethodNameTransform = mn => Invariant($"$/invokeProxy/{handle}/{attribute.SubTypeCode}.{context.JsonRpcTargetOptions.MethodNameTransform?.Invoke(mn) ?? mn}"),
```

Comment:
This is tested differently:
1. on the server side, method ordering is not important because we don't have name collisions: we use [TypeInfo.DeclaredMethods](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.typeinfo.declaredmethods) to retrieve the list of methods for each interface. `DeclaredMethods` doesn't return hidden methods (not clearly documented, but I tested it to be sure). Methods that have the same name across different known interfaces are registered as different RPC names with the `n.foo` pattern.
2. on the client side, the order is important. That is tested by making sure that various combinations of interface implementations result in a dispatch to the right method.

Do you want something else?

Reply:
What I mean is, do we have tests where a custom method name transform is applied to a marshalable object's interface? I think there isn't. Yet we have functionality here that seems to support it, so I'm thinking we ought to test that.

Reply:
You cannot apply a method name transform to a marshaled object's method.

I added a test to validate that a method name transform on the server doesn't trickle down to the marshaled object methods. And also to verify that applying a method name transform on the server object doesn't break the marshaled object functionality.